File: backend-fastapi/app/routes/admin_testimonials.py
"""
Routes d'administration pour les témoignages
"""

from datetime import datetime, timedelta
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from sqlalchemy import func, desc, asc

from app.database import get_db
from app.models import Admin, Testimonial
from app.schemas_testimonials import (
    TestimonialCreate, 
    TestimonialUpdate, 
    TestimonialResponse, 
    TestimonialStats
)
from app.security import get_current_admin

router = APIRouter()

@router.get("/", response_model=List[TestimonialResponse])
async def get_all_testimonials(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=100),
    is_active: Optional[bool] = None,
    search: Optional[str] = None,
    sort_by: Optional[str] = Query("created_at", regex="^(created_at|nom|ecole|rating)$"),
    sort_order: Optional[str] = Query("desc", regex="^(asc|desc)$"),
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin)
):
    """
    Récupérer tous les témoignages avec filtres et pagination
    """
    query = db.query(Testimonial)
    
    # Filtre par statut actif/inactif
    if is_active is not None:
        query = query.filter(Testimonial.is_active == is_active)
    
    # Recherche par nom, école ou commentaire
    if search:
        search_term = f"%{search}%"
        query = query.filter(
            Testimonial.nom.ilike(search_term) |
            Testimonial.ecole.ilike(search_term) |
            Testimonial.comment.ilike(search_term)
        )
    
    # Tri
    if sort_order == "desc":
        query = query.order_by(desc(getattr(Testimonial, sort_by)))
    else:
        query = query.order_by(asc(getattr(Testimonial, sort_by)))
    
    # Pagination
    testimonials = query.offset(skip).limit(limit).all()
    
    return testimonials

@router.get("/stats", response_model=TestimonialStats)
async def get_testimonial_stats(
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin)
):
    """
    Obtenir les statistiques des témoignages
    """
    total_testimonials = db.query(Testimonial).count()
    active_testimonials = db.query(Testimonial).filter(Testimonial.is_active == True).count()
    
    # Moyenne des notes
    avg_rating = db.query(func.avg(Testimonial.rating)).filter(Testimonial.is_active == True).scalar()
    avg_rating = round(float(avg_rating or 0), 2)
    
    # Témoignages récents (30 derniers jours)
    thirty_days_ago = datetime.utcnow() - timedelta(days=30)
    recent_testimonials = db.query(Testimonial).filter(
        Testimonial.created_at >= thirty_days_ago
    ).count()
    
    return TestimonialStats(
        total_testimonials=total_testimonials,
        active_testimonials=active_testimonials,
        average_rating=avg_rating,
        recent_testimonials=recent_testimonials
    )

@router.get("/{testimonial_id}", response_model=TestimonialResponse)
async def get_testimonial(
    testimonial_id: int,
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin)
):
    """
    Récupérer un témoignage spécifique
    """
    testimonial = db.query(Testimonial).filter(Testimonial.id == testimonial_id).first()
    if not testimonial:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Témoignage non trouvé"
        )
    return testimonial


@router.put("/{testimonial_id}", response_model=TestimonialResponse)
async def update_testimonial(
    testimonial_id: int,
    testimonial_data: TestimonialUpdate,
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin)
):
    """
    Modifier un témoignage existant
    """
    testimonial = db.query(Testimonial).filter(Testimonial.id == testimonial_id).first()
    if not testimonial:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Témoignage non trouvé"
        )
    
    try:
        # Mise à jour des champs modifiés uniquement
        update_data = testimonial_data.dict(exclude_unset=True)
        for field, value in update_data.items():
            setattr(testimonial, field, value)
        
        testimonial.updated_at = datetime.utcnow()
        
        db.commit()
        db.refresh(testimonial)
        
        return testimonial
        
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Erreur lors de la modification du témoignage: {str(e)}"
        )

@router.delete("/{testimonial_id}")
async def delete_testimonial(
    testimonial_id: int,
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin)
):
    """
    Supprimer un témoignage
    """
    testimonial = db.query(Testimonial).filter(Testimonial.id == testimonial_id).first()
    if not testimonial:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Témoignage non trouvé"
        )
    
    try:
        db.delete(testimonial)
        db.commit()
        
        return {"message": f"Témoignage '{testimonial.nom}' supprimé avec succès"}
        
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Erreur lors de la suppression du témoignage: {str(e)}"
        )

@router.put("/{testimonial_id}/toggle-status")
async def toggle_testimonial_status(
    testimonial_id: int,
    db: Session = Depends(get_db),
    current_admin: Admin = Depends(get_current_admin)
):
    """
    Activer/désactiver un témoignage
    """
    testimonial = db.query(Testimonial).filter(Testimonial.id == testimonial_id).first()
    if not testimonial:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Témoignage non trouvé"
        )
    
    try:
        testimonial.is_active = not testimonial.is_active
        testimonial.updated_at = datetime.utcnow()
        
        db.commit()
        db.refresh(testimonial)
        
        status_text = "activé" if testimonial.is_active else "désactivé"
        return {"message": f"Témoignage '{testimonial.nom}' {status_text} avec succès"}
        
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Erreur lors de la modification du statut: {str(e)}"
        )

File: frontend/src/pages/admin/AdminTestimonials.js
import React, { useState, useEffect } from "react"
import { useNavigate } from "react-router-dom"
import AdminLayout from "../../components/admin/AdminLayout"
import { 
  getAllTestimonials, 
  updateTestimonial, 
  deleteTestimonial, 
  toggleTestimonialStatus,
  getTestimonialStats
} from "../../utils/api"
import "./AdminTestimonials.css"

const AdminTestimonials = () => {
  const [testimonials, setTestimonials] = useState([])
  const [filteredTestimonials, setFilteredTestimonials] = useState([])
  const [stats, setStats] = useState({})
  const [filters, setFilters] = useState({
    status: "all",
    search: "",
    sortBy: "created_at",
    sortOrder: "desc"
  })
  const [isLoading, setIsLoading] = useState(true)
  const [editingTestimonial, setEditingTestimonial] = useState(null)
  const [editForm, setEditForm] = useState({
    nom: "",
    ecole: "", 
    comment: "",
    rating: 5
  })
  const navigate = useNavigate()

  useEffect(() => {
    fetchTestimonials()
    fetchStats()
  }, [filters.sortBy, filters.sortOrder])

  useEffect(() => {
    applyFilters()
  }, [testimonials, filters.status, filters.search])

  const fetchTestimonials = async () => {
    try {
      const params = {
        sort_by: filters.sortBy,
        sort_order: filters.sortOrder,
        limit: 100
      }
      const data = await getAllTestimonials(params)
      setTestimonials(data)
    } catch (error) {
      console.error("Erreur lors du chargement des témoignages:", error)
    } finally {
      setIsLoading(false)
    }
  }

  const fetchStats = async () => {
    try {
      const data = await getTestimonialStats()
      setStats(data)
    } catch (error) {
      console.error("Erreur lors du chargement des statistiques:", error)
    }
  }

  const applyFilters = () => {
    let filtered = testimonials

    if (filters.status !== "all") {
      filtered = filtered.filter(t => 
        filters.status === "active" ? t.is_active : !t.is_active
      )
    }

    if (filters.search) {
      const searchLower = filters.search.toLowerCase()
      filtered = filtered.filter(t =>
        t.nom.toLowerCase().includes(searchLower) ||
        t.ecole.toLowerCase().includes(searchLower) ||
        t.comment.toLowerCase().includes(searchLower)
      )
    }

    setFilteredTestimonials(filtered)
  }

  const handleFilterChange = (key, value) => {
    setFilters(prev => ({ ...prev, [key]: value }))
  }

  const handleEdit = (testimonial) => {
    setEditingTestimonial(testimonial.id)
    setEditForm({
      nom: testimonial.nom,
      ecole: testimonial.ecole,
      comment: testimonial.comment,
      rating: testimonial.rating
    })
  }

  const handleEditSubmit = async (e) => {
    e.preventDefault()
    try {
      await updateTestimonial(editingTestimonial, editForm)
      await fetchTestimonials()
      setEditingTestimonial(null)
      setEditForm({ nom: "", ecole: "", comment: "", rating: 5 })
      alert("Témoignage modifié avec succès !")
    } catch (error) {
      alert("Erreur lors de la modification")
      console.error(error)
    }
  }

  const handleDelete = async (id, name) => {
    if (window.confirm(`Êtes-vous sûr de vouloir supprimer le témoignage de "${name}" ?`)) {
      try {
        await deleteTestimonial(id)
        await fetchTestimonials()
        await fetchStats()
        alert("Témoignage supprimé avec succès !")
      } catch (error) {
        alert("Erreur lors de la suppression")
        console.error(error)
      }
    }
  }

  const handleToggleStatus = async (id) => {
    try {
      await toggleTestimonialStatus(id)
      await fetchTestimonials()
      await fetchStats()
      alert("Statut modifié avec succès !")
    } catch (error) {
      alert("Erreur lors du changement de statut")
      console.error(error)
    }
  }

  const formatDate = (dateString) => {
    return new Date(dateString).toLocaleDateString("fr-FR", {
      year: "numeric",
      month: "long", 
      day: "numeric",
      hour: "2-digit",
      minute: "2-digit"
    })
  }

  if (isLoading) {
    return (
      <AdminLayout>
        <div className="admin-testimonials-loading-modern">
          <div className="loading-spinner-testimonials-modern"></div>
          <div className="loading-dots-testimonials-modern">
            <span></span>
            <span></span>
            <span></span>
          </div>
          <p>Chargement des témoignages...</p>
        </div>
      </AdminLayout>
    )
  }

  return (
    <AdminLayout>
      <div className="admin-testimonials-page-modern">
        {/* Header moderne avec statistiques */}
        <div className="page-header-testimonials-modern">
          <div className="header-content-testimonials-modern">
            <div className="header-icon-testimonials-modern">💬</div>
            <div className="header-text-testimonials-modern">
              <h1>Gestion des Témoignages</h1>
              <p>Gérez les avis et témoignages de vos étudiants</p>
            </div>
          </div>
        </div>

        {/* Statistiques modernes */}
        <div className="stats-section-testimonials-modern">
          <div className="stat-card-testimonials-modern total">
            <div className="stat-icon-testimonials-modern">📊</div>
            <div className="stat-content-testimonials-modern">
              <div className="stat-number-testimonials-modern">{stats.total_testimonials || 0}</div>
              <div className="stat-label-testimonials-modern">Total</div>
            </div>
          </div>
          <div className="stat-card-testimonials-modern active">
            <div className="stat-icon-testimonials-modern">✅</div>
            <div className="stat-content-testimonials-modern">
              <div className="stat-number-testimonials-modern">{stats.active_testimonials || 0}</div>
              <div className="stat-label-testimonials-modern">Actifs</div>
            </div>
          </div>
          <div className="stat-card-testimonials-modern rating">
            <div className="stat-icon-testimonials-modern">⭐</div>
            <div className="stat-content-testimonials-modern">
              <div className="stat-number-testimonials-modern">{stats.average_rating || 0}</div>
              <div className="stat-label-testimonials-modern">Note moyenne</div>
            </div>
          </div>
          <div className="stat-card-testimonials-modern recent">
            <div className="stat-icon-testimonials-modern">📅</div>
            <div className="stat-content-testimonials-modern">
              <div className="stat-number-testimonials-modern">{stats.recent_testimonials || 0}</div>
              <div className="stat-label-testimonials-modern">Ce mois</div>
            </div>
          </div>
        </div>

        {/* Filtres modernes */}
        <div className="testimonials-filters-modern">
          <div className="filters-container-modern">
            <div className="search-field-testimonials-modern">
              <span className="search-icon-testimonials-modern">🔍</span>
              <input
                type="text"
                placeholder="Rechercher par nom, école ou commentaire..."
                value={filters.search}
                onChange={(e) => handleFilterChange("search", e.target.value)}
                className="search-input-testimonials-modern"
              />
            </div>
            
            <select
              value={filters.status}
              onChange={(e) => handleFilterChange("status", e.target.value)}
              className="filter-select-testimonials-modern"
            >
              <option value="all">Tous les témoignages</option>
              <option value="active">Actifs seulement</option>
              <option value="inactive">Inactifs seulement</option>
            </select>

            <select
              value={`${filters.sortBy}-${filters.sortOrder}`}
              onChange={(e) => {
                const [sortBy, sortOrder] = e.target.value.split("-")
                handleFilterChange("sortBy", sortBy)
                handleFilterChange("sortOrder", sortOrder)
              }}
              className="filter-select-testimonials-modern"
            >
              <option value="created_at-desc">Plus récents</option>
              <option value="created_at-asc">Plus anciens</option>
              <option value="rating-desc">Meilleures notes</option>
              <option value="nom-asc">Nom A-Z</option>
            </select>
          </div>
        </div>

        {/* Section des témoignages */}
        <div className="testimonials-section-modern">
          <div className="section-header-testimonials-modern">
            <h2>Témoignages ({filteredTestimonials.length})</h2>
            <p>Gérez et modérez les avis de vos étudiants</p>
          </div>

          {filteredTestimonials.length === 0 ? (
            <div className="empty-state-testimonials-modern">
              <div className="empty-illustration-testimonials-modern">
                <div className="empty-icon-testimonials-modern">💬</div>
                <div className="empty-graphics-testimonials-modern"></div>
              </div>
              <div className="empty-content-testimonials-modern">
                <h3>
                  {filters.search || filters.status !== "all" 
                    ? "Aucun témoignage trouvé" 
                    : "Aucun témoignage disponible"}
                </h3>
                <p>
                  {filters.search || filters.status !== "all" 
                    ? "Aucun témoignage ne correspond à vos critères de recherche"
                    : "Les témoignages soumis par les utilisateurs apparaîtront ici"}
                </p>
              </div>
            </div>
          ) : (
            <div className="testimonials-grid-modern">
              {filteredTestimonials.map((testimonial, index) => (
                <div 
                  key={testimonial.id} 
                  className={`testimonial-card-modern ${!testimonial.is_active ? "inactive" : ""}`}
                  style={{ animationDelay: `${index * 0.1}s` }}
                >
                  {editingTestimonial === testimonial.id ? (
                    <div className="edit-form-modern">
                      <div className="edit-header-modern">
                        <h4>✏️ Modifier le témoignage</h4>
                      </div>
                      <form onSubmit={handleEditSubmit}>
                        <div className="form-grid-testimonials-modern">
                          <input
                            type="text"
                            placeholder="Nom"
                            value={editForm.nom}
                            onChange={(e) => setEditForm({...editForm, nom: e.target.value})}
                            required
                            className="form-input-testimonials-modern"
                          />
                          <input
                            type="text"
                            placeholder="École"
                            value={editForm.ecole}
                            onChange={(e) => setEditForm({...editForm, ecole: e.target.value})}
                            required
                            className="form-input-testimonials-modern"
                          />
                        </div>
                        <textarea
                          placeholder="Commentaire"
                          value={editForm.comment}
                          onChange={(e) => setEditForm({...editForm, comment: e.target.value})}
                          required
                          className="form-textarea-testimonials-modern"
                        />
                        <div className="form-actions-testimonials-modern">
                          <button type="submit" className="btn-save-testimonials-modern">
                            💾 Sauvegarder
                          </button>
                          <button 
                            type="button" 
                            className="btn-cancel-testimonials-modern"
                            onClick={() => setEditingTestimonial(null)}
                          >
                            Annuler
                          </button>
                        </div>
                      </form>
                    </div>
                  ) : (
                    <div className="testimonial-content-modern">
                      <div className="testimonial-header-modern">
                        <div className="author-info-modern">
                          <h3 className="author-name-modern">{testimonial.nom}</h3>
                          <p className="author-school-modern">🎓 {testimonial.ecole}</p>
                        </div>
                        <div className="testimonial-badges-modern">
                          <div className="rating-modern">
                            {Array.from({ length: testimonial.rating }).map((_, i) => (
                              <span key={i} className="star-modern">⭐</span>
                            ))}
                          </div>
                          <span className={`status-badge-testimonials-modern ${testimonial.is_active ? "active" : "inactive"}`}>
                            {testimonial.is_active ? "✅ Actif" : "❌ Inactif"}
                          </span>
                        </div>
                      </div>
                      
                      <div className="comment-modern">
                        <span className="quote-icon-modern">💭</span>
                        {testimonial.comment}
                      </div>
                      
                      <div className="testimonial-meta-modern">
                        <span className="date-modern">📅 {formatDate(testimonial.created_at)}</span>
                      </div>

                      <div className="testimonial-actions-modern">
                        <button
                          className="btn-action-testimonials-modern edit"
                          onClick={() => handleEdit(testimonial)}
                          title="Modifier"
                        >
                          ✏️
                        </button>
                        <button
                          className={`btn-action-testimonials-modern ${testimonial.is_active ? "deactivate" : "activate"}`}
                          onClick={() => handleToggleStatus(testimonial.id)}
                          title={testimonial.is_active ? "Désactiver" : "Activer"}
                        >
                          {testimonial.is_active ? "👁️‍🗨️" : "👁️"}
                        </button>
                        <button
                          className="btn-action-testimonials-modern delete"
                          onClick={() => handleDelete(testimonial.id, testimonial.nom)}
                          title="Supprimer"
                        >
                          🗑️
                        </button>
                      </div>
                    </div>
                  )}
                </div>
              ))}
            </div>
          )}
        </div>
      </div>
    </AdminLayout>
  )
}

export default AdminTestimonials
